placement/placement.go

```diff
@@ -0,0 +1,300 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package placement
+
+import (
+	"encoding/json"
+	"errors"
+	"sort"
+)
+
+var (
+	errShardsWithDifferentReplicas = errors.New("invalid placement, found shards with different number of replications")
+	errInvalidHostShards           = errors.New("invalid shards assigned to a host")
+)
+
+// snapshot implements Snapshot
+type snapshot struct {
+	hostShards []HostShards
+	rf         int
+	shards     []uint32
+}
+
+// NewEmptyPlacementSnapshot returns an empty placement
+func NewEmptyPlacementSnapshot(hosts []Host, ids []uint32) Snapshot {
+	hostShards := make([]HostShards, len(hosts), len(hosts))
+	for i, ph := range hosts {
+		hostShards[i] = NewEmptyHostShardsFromHost(ph)
+	}
+
+	return snapshot{hostShards: hostShards, shards: ids, rf: 0}
+}
+
+// NewPlacementSnapshot returns a placement
+func NewPlacementSnapshot(hss []HostShards, shards []uint32, rf int) Snapshot {
+	return snapshot{hostShards: hss, rf: rf, shards: shards}
+}
+
+func (ps snapshot) HostShards() []HostShards {
+	result := make([]HostShards, ps.HostsLen())
+	for i, hs := range ps.hostShards {
+		result[i] = hs
+	}
+	return result
+}
+
+func (ps snapshot) HostsLen() int {
+	return len(ps.hostShards)
+}
+
+func (ps snapshot) Replicas() int {
+	return ps.rf
+}
+
+func (ps snapshot) ShardsLen() int {
+	return len(ps.shards)
+}
+
+func (ps snapshot) Shards() []uint32 {
+	return ps.shards
+}
+
+func (ps snapshot) HostShard(id string) HostShards {
+	for _, phs := range ps.HostShards() {
+		if phs.Host().ID() == id {
+			return phs
+		}
+	}
+	return nil
+}
+
+func (ps snapshot) Validate() bool {
+	set := getUniqueIDSet(ps.shards)
+	if len(set) != len(ps.shards) {
+		return false
+	}
+
+	expectedTotal := len(ps.shards) * ps.rf
+	actualTotal := 0
+	for _, hs := range ps.hostShards {
+		for _, id := range hs.Shards() {
+			if _, exist := set[id]; !exist {
+				return false
+			}
+		}
+		actualTotal += hs.ShardsLen()
+	}
+
+	if expectedTotal != actualTotal {
+		return false
+	}
+	return true
+}
+
+func getUniqueIDSet(ids []uint32) map[uint32]struct{} {
```

Comment:
hm.. I don't like having two methods that are exactly the same. Can we put the implementation somewhere that can be called from different packages?